File: app/src/main/java/app/amber/feature/miniapp/MiniAppHtmlValidator.kt
package app.amber.feature.miniapp

class MiniAppValidationException(message: String) : IllegalArgumentException(message)

object MiniAppHtmlValidator {
    private val requiredHtmlPattern = Regex("""(?is)<\s*(html\b|!doctype\s+html)""")
    private val blockedPatterns = listOf(
        Regex("""(?is)<\s*script\b[^>]*\bsrc\s*=""") to "External scripts are not allowed",
        Regex("""(?is)<\s*(iframe|object|embed|form)\b""") to "Embedded/submit-capable elements are not allowed",
        Regex("""(?is)<\s*(img|source)\b[^>]*\bsrcset\s*=""") to "Image srcset is not supported",
        Regex("""(?is)<\s*link\b[^>]*\bhref\s*=\s*['"]?\s*https?:""") to "External stylesheets are not allowed",
        Regex("""(?is)@import\s+(['"]?\s*)?(https?:|//|file:|content:)""") to "CSS imports are not allowed",
        Regex("""(?is)url\s*\(\s*(['"]?)\s*(https?:|//|file:|content:)""") to "External CSS URLs are not allowed",
        Regex("""(?is)\beval\s*\(""") to "eval() is not allowed",
        Regex("""(?is)\bnew\s+Function\b""") to "new Function is not allowed",
        Regex("""(?is)\bimport\s*\(""") to "dynamic import() is not allowed",
        Regex("""(?is)\bimport\s+['"]""") to "static import is not allowed",
        Regex("""(?is)\bWebSocket\b""") to "WebSocket is not allowed",
        Regex("""(?is)\bEventSource\b""") to "EventSource is not allowed",
        Regex("""(?is)\bXMLHttpRequest\b""") to "XMLHttpRequest is not allowed",
        Regex("""(?is)\blocalStorage\b""") to "localStorage is not allowed",
        Regex("""(?is)\bsessionStorage\b""") to "sessionStorage is not allowed",
        Regex("""(?is)\bindexedDB\b""") to "indexedDB is not allowed",
        Regex("""(?is)\bnavigator\s*\.\s*geolocation\b""") to "geolocation is not allowed",
        Regex("""(?is)\bnavigator\s*\.\s*mediaDevices\b""") to "mediaDevices is not allowed",
        Regex("""(?is)\bnavigator\s*\.\s*clipboard\b""") to "native clipboard is not allowed",
        Regex("""(?is)\bnavigator\s*\[\s*['"]\s*(geolocation|mediaDevices|clipboard)\s*['"]\s*\]""") to "computed access to blocked navigator APIs is not allowed",
        Regex("""(?is)\bwindow\s*\[\s*['"]\s*(XMLHttpRequest|WebSocket|EventSource|localStorage|sessionStorage|indexedDB)\s*['"]\s*\]""") to "computed access to blocked browser APIs is not allowed",
        Regex("""(?is)\bglobalThis\s*\[\s*['"]\s*(XMLHttpRequest|WebSocket|EventSource|localStorage|sessionStorage|indexedDB)\s*['"]\s*\]""") to "computed access to blocked browser APIs is not allowed",
    )
    private val quotedImageResourcePattern =
        Regex("""(?is)<\s*(img|source)\b[^>]*\b(src|srcset)\s*=\s*(['"])(.*?)\3""")
    private val unquotedImageResourcePattern =
        Regex("""(?is)<\s*(img|source)\b[^>]*\b(src|srcset)\s*=\s*([^\s"'=<>`]+)""")

    fun validate(html: String) {
        val sizeBytes = html.encodeToByteArray().size
        if (sizeBytes > MINI_APP_MAX_HTML_BYTES) {
            throw MiniAppValidationException("HTML is too large: $sizeBytes bytes")
        }
        if (!requiredHtmlPattern.containsMatchIn(html)) {
            throw MiniAppValidationException("HTML must include <html> or <!DOCTYPE html>")
        }
        blockedPatterns.firstOrNull { (pattern, _) -> pattern.containsMatchIn(html) }?.let { (_, reason) ->
            throw MiniAppValidationException(reason)
        }
        if (hasInvalidImageResource(html)) {
            throw MiniAppValidationException("MiniApp images must use data:image or https URLs")
        }
    }

    private fun hasInvalidImageResource(html: String): Boolean {
        val quotedInvalid = quotedImageResourcePattern.findAll(html).any { match ->
            !match.groupValues[4].trim().isAllowedImageUrl()
        }
        if (quotedInvalid) return true
        return unquotedImageResourcePattern.findAll(html).any { match ->
            !match.groupValues[3].trim().isAllowedImageUrl()
        }
    }

    private fun String.isAllowedImageUrl(): Boolean {
        return startsWith("data:image/", ignoreCase = true) || startsWith("https://", ignoreCase = true)
    }
}

File: iosApp/iosApp/MiniAppHtmlValidator.swift
import Foundation

/// [MiniApp MVP] iOS port of Android's `MiniAppHtmlValidator`
/// (app/src/main/java/app/amber/feature/miniapp/MiniAppHtmlValidator.kt).
///
/// Validates generated MiniApp HTML before it is loaded into the WKWebView
/// runner. Same rule set as Android so the security posture is equivalent:
/// reject external scripts, non-https images, dangerous browser APIs (eval,
/// dynamic import, XHR, WebSocket, storage, geo/media/clipboard), and
/// iframe/object/embed/form elements. Allow inline `<script>` (the MiniApp's
/// own logic) calling the Amber bridge API.
///
/// This is the load-bearing security gate for the MiniApp runner — without it,
/// generated HTML could exfiltrate data or reach the network directly.
enum MiniAppHtmlValidator {

    /// Max HTML size, matching Android's MINI_APP_MAX_HTML_BYTES = 768 * 1024
    /// (app/src/main/java/app/amber/feature/miniapp/MiniAppModels.kt:6).
    static let maxHtmlBytes = 768 * 1024

    struct ValidationError: Error, LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    // All patterns are case-insensitive + dotall (match newlines), matching the
    // Kotlin `(?is)` flag. NSRegularExpression options: [.caseInsensitive, .dotMatchesLineSeparators].
    private struct BlockedRule {
        let pattern: String
        let reason: String
    }

    private static let requiredHtmlPattern = #"<\s*(html\b|!doctype\s+html)"#
    private static let blockedRules: [BlockedRule] = [
        .init(pattern: ##"<\s*script\b[^>]*\bsrc\s*="##, reason: "External scripts are not allowed"),
        .init(pattern: #"<\s*(iframe|object|embed|form)\b"#, reason: "Embedded/submit-capable elements are not allowed"),
        .init(pattern: ##"<\s*(img|source)\b[^>]*\bsrcset\s*="##, reason: "Image srcset is not supported"),
        .init(pattern: #"<\s*link\b[^>]*\bhref\s*=\s*['"]?\s*https?:"#, reason: "External stylesheets are not allowed"),
        .init(pattern: #"@import\s+(['"]?\s*)?(https?:|//|file:|content:)"#, reason: "CSS imports are not allowed"),
        .init(pattern: #"url\s*\(\s*(['"]?)\s*(https?:|//|file:|content:)"#, reason: "External CSS URLs are not allowed"),
        .init(pattern: #"\beval\s*\("#, reason: "eval() is not allowed"),
        .init(pattern: #"\bnew\s+Function\b"#, reason: "new Function is not allowed"),
        .init(pattern: #"\bimport\s*\("#, reason: "dynamic import() is not allowed"),
        .init(pattern: #"\bimport\s+['"]"#, reason: "static import is not allowed"),
        .init(pattern: #"\bWebSocket\b"#, reason: "WebSocket is not allowed"),
        .init(pattern: #"\bEventSource\b"#, reason: "EventSource is not allowed"),
        .init(pattern: #"\bXMLHttpRequest\b"#, reason: "XMLHttpRequest is not allowed"),
        .init(pattern: #"\blocalStorage\b"#, reason: "localStorage is not allowed"),
        .init(pattern: #"\bsessionStorage\b"#, reason: "sessionStorage is not allowed"),
        .init(pattern: #"\bindexedDB\b"#, reason: "indexedDB is not allowed"),
        .init(pattern: #"\bnavigator\s*\.\s*geolocation\b"#, reason: "geolocation is not allowed"),
        .init(pattern: #"\bnavigator\s*\.\s*mediaDevices\b"#, reason: "mediaDevices is not allowed"),
        .init(pattern: #"\bnavigator\s*\.\s*clipboard\b"#, reason: "native clipboard is not allowed"),
        .init(pattern: #"\bnavigator\s*\[\s*['"]\s*(geolocation|mediaDevices|clipboard)\s*['"]\s*\]"#, reason: "computed access to blocked navigator APIs is not allowed"),
        .init(pattern: #"\bwindow\s*\[\s*['"]\s*(XMLHttpRequest|WebSocket|EventSource|localStorage|sessionStorage|indexedDB)\s*['"]\s*\]"#, reason: "computed access to blocked browser APIs is not allowed"),
        .init(pattern: #"\bglobalThis\s*\[\s*['"]\s*(XMLHttpRequest|WebSocket|EventSource|localStorage|sessionStorage|indexedDB)\s*['"]\s*\]"#, reason: "computed access to blocked browser APIs is not allowed"),
    ]

    // Image resource extraction (src/srcset), quoted + unquoted.
    private static let quotedImageResourcePattern = ##"<\s*(img|source)\b[^>]*\b(src|srcset)\s*=\s*(['"])(.*?)\3"##
    private static let unquotedImageResourcePattern = ##"<\s*(img|source)\b[^>]*\b(src|srcset)\s*=\s*([^\s"'=<>`]+)"##

    /// Validate `html`. Throws `ValidationError` if any rule is violated.
    /// Mirrors Android's `MiniAppHtmlValidator.validate` exactly.
    static func validate(_ html: String) throws {
        let sizeBytes = html.utf8.count
        if sizeBytes > maxHtmlBytes {
            throw ValidationError(message: "HTML is too large: \(sizeBytes) bytes")
        }
        let opts: NSRegularExpression.Options = [.caseInsensitive, .dotMatchesLineSeparators]
        let fullRange = NSRange(html.startIndex..., in: html)

        if try regex(requiredHtmlPattern, opts).firstMatch(in: html, range: fullRange) == nil {
            throw ValidationError(message: "HTML must include <html> or <!DOCTYPE html>")
        }
        for rule in blockedRules {
            if try regex(rule.pattern, opts).firstMatch(in: html, range: fullRange) != nil {
                throw ValidationError(message: rule.reason)
            }
        }
        if try hasInvalidImageResource(html, opts: opts) {
            throw ValidationError(message: "MiniApp images must use data:image or https URLs")
        }
    }

    private static func regex(_ pattern: String, _ opts: NSRegularExpression.Options) throws -> NSRegularExpression {
        try NSRegularExpression(pattern: pattern, options: opts)
    }

    private static func hasInvalidImageResource(_ html: String, opts: NSRegularExpression.Options) throws -> Bool {
        let fullRange = NSRange(html.startIndex..., in: html)
        // Quoted: capture group 4 is the URL.
        let quoted = try regex(quotedImageResourcePattern, opts)
        for m in quoted.matches(in: html, range: fullRange) {
            if let r = Range(m.range(at: 4), in: html),
               !html[r].trimmingCharacters(in: .whitespaces).isAllowedImageUrl {
                return true
            }
        }
        // Unquoted: capture group 3 is the URL.
        let unquoted = try regex(unquotedImageResourcePattern, opts)
        for m in unquoted.matches(in: html, range: fullRange) {
            if let r = Range(m.range(at: 3), in: html),
               !html[r].trimmingCharacters(in: .whitespaces).isAllowedImageUrl {
                return true
            }
        }
        return false
    }
}

private extension String {
    var isAllowedImageUrl: Bool {
        lowercased().hasPrefix("data:image/") || lowercased().hasPrefix("https://")
    }
}

